flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/StopWithSavepointContextTest.java

```diff
@@ -0,0 +1,249 @@
+import static org.junit.Assert.fail;
+
+/**
+ * {@code StopWithSavepointContextTest} tests the stop-with-savepoint functionality of {@link
+ * SchedulerBase#stopWithSavepoint(String, boolean)}.
+ */
+public class StopWithSavepointContextTest extends TestLogger {
+
+    private JobGraph jobGraph;
+    private DefaultScheduler scheduler;
+
+    private StopWithSavepointOperations testInstance;
+
+    @Before
+    public void setup() throws Exception {
+        jobGraph = new JobGraph();
+
+        final JobVertex jobVertex = new JobVertex("vertex #0");
+        jobVertex.setInvokableClass(NoOpInvokable.class);
+        jobGraph.addVertex(jobVertex);
+
+        // checkpointInterval has to be set to a value lower than Long.MAX_VALUE to enable
+        // periodic checkpointing - only then can we enable/disable the CheckpointCoordinator
+        SchedulerTestingUtils.enablePeriodicCheckpointing(jobGraph, Long.MAX_VALUE - 1);
+        scheduler =
+                SchedulerTestingUtils.createSchedulerBuilder(
+                                jobGraph, ComponentMainThreadExecutorServiceAdapter.forMainThread())
+                        .setFutureExecutor(new DirectScheduledExecutorService())
+                        .build();
+        scheduler.startScheduling();
+
+        // the checkpoint scheduler is stopped before triggering the stop-with-savepoint
+        disableCheckpointScheduler();
+
+        testInstance = new StopWithSavepointContext(jobGraph.getJobID(), scheduler, this.log);
+    }
+
+    @Test
+    public void testHappyPathWithSavepointCreationBeforeTermination() throws Exception {
+        assertHappyPath(
+                (savepointPath) -> {
+                    testInstance.handleSavepointCreation(savepointPath, null);
+                    testInstance.handleExecutionTermination(
+                            Collections.singletonList(ExecutionState.FINISHED));
+                });
+    }
+
+    @Test
+    public void testHappyPathWithSavepointCreationAfterTermination() throws Exception {
+        assertHappyPath(
+                (savepointPath) -> {
+                    testInstance.handleExecutionTermination(
+                            Collections.singletonList(ExecutionState.FINISHED));
+                    testInstance.handleSavepointCreation(savepointPath, null);
+                });
+    }
+
+    private void assertHappyPath(Consumer<String> stopWithSavepointCompletion) throws Exception {
+        final String savepointPath = "savepoint-path";
+
+        stopWithSavepointCompletion.accept(savepointPath);
+
+        assertThat(testInstance.getResult().get(), is(savepointPath));
+
+        // the happy path won't restart the CheckpointCoordinator
+        assertCheckpointSchedulingDisabled();
+    }
+
+    @Test
+    public void testSavepointCreationFailure() {
+        final Exception exception = new Exception("Expected exception during savepoint creation.");
+        testInstance.handleSavepointCreation(null, exception);
```

Comment:
What happens if `testInstance.onExecutionsTermination` is called after the savepoint has failed (e.g. if an unrelated global failover is triggered at a later point in time)?

Reply:
This test case is covered in `StopWithSavepointTerminationHandlerImplTest.testSavepointCreationFailureWithFailingExecutions`.